### PR TITLE
Move page and page_size query param validation into serializer

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Related to #857 and #843 

## Description

In this PR I move scattered query parameter validation into the query serializer, where it probably belongs. This has two benefits:

1. It removes code in favour of using built-in Django Rest Framework serializer validation
2. It removes usages of the bespoke `get_api_exception`
3. It reduces instances of a loose pattern of scattered validation where parameter validation happens multiple times and deep into the runtime of a search query, rather than once at the top (where it is expected to happen and easier to find)

None of these changes change the API surface.

The lines changed is also deceptive on this PR as something like 85 of the lines added or changed were in the tests.

## Testing Instructions

Ensure CI passes. Checkout the branch, set up the local API following the README instructions and make some smoke test requests to ensure everything works as expected.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: api/catalog/api/controllers/search_controller.py [hunk 1]
```diff
@@ -425,10 +424,9 @@ def _get_result_and_page_count(
     :return: Result and page count.
     """
     result_count = response_obj.hits.total.value
-    natural_page_count = int(result_count / page_size)
-    if natural_page_count % page_size != 0:
-        natural_page_count += 1
-    page_count = min(natural_page_count, MAX_TOTAL_PAGE_COUNT)
```
Comment (sarayourfriend) [inline api/catalog/api/controllers/search_controller.py:431]: This clamping is moved up into the paginator and out of the search logic.

**Comment by sarayourfriend** [comment]
Keeping this PR as a draft as it should not be merged until we've fixed the production ES CPU usage.

**Comment by dhruvkb** [comment]
@sarayourfriend since the ES CPU usage is under control now, should we push to update, undraft and get this PR merged?

**Comment by dhruvkb** [comment]
Diff before updating to `main`, preserved for self-reference.

image

File: api/catalog/api/serializers/media_serializers.py [hunk 2]
```diff
@@ -161,10 +175,30 @@ def validate_title(self, value):
...
+            if is_anonymous
+            else settings.MAX_AUTHED_PAGE_SIZE
+        )
+
+        validator = MaxValueValidator(
+            max_value,
+            message=serializers.IntegerField.default_error_messages["max_value"].format(
+                max_value=max_value
+            ),
+        )
```
Comment (dhruvkb) [inline api/catalog/api/serializers/media_serializers.py:189]: Can be extracted into a separate validator, and applied to the field using the [`validators` argument](https://www.django-rest-framework.org/api-guide/fields/#validators). That removes the need for the `try`-`except` handling below.

File: api/test/unit/controllers/test_search_controller.py [hunk 3]
```diff
@@ -41,8 +40,7 @@
         (20, 5, 5, (20, 5)),
         # Fewer hits than page size, but result list somehow differs, use that for count
         (48, 20, 50, (20, 0)),
-        # Page count gets truncated always
-        (5000, 10, 10, (5000, MAX_TOTAL_PAGE_COUNT)),
+        (5000, 10, 10, (5000, 5000 / 10)),
```
Comment (dhruvkb) [inline api/test/unit/controllers/test_search_controller.py:43]: Need a comment here, explaining what this test case is testing for.

**Review by dhruvkb — APPROVED** [review]

**Review by dhruvkb — APPROVED** [review]
This can be compared against an earlier (approved) implementation in #778, so this PR should be too!

File: api/test/unit/controllers/test_search_controller.py (hunk 3 above)
Comment (krysal) [reply api/test/unit/controllers/test_search_controller.py:43]: I believe we can remove this case, this function does not truncate the `page_count` anymore as that logic relies on the serializer now.

File: api/catalog/api/serializers/media_serializers.py (hunk 2 above)
Comment (krysal) [reply api/catalog/api/serializers/media_serializers.py:189]: Great! We can do that on another refactor to provide a more specific error. Currently, it says `"Ensure this value is less than or equal to 20."` when before it was hinted that that was just the limit for unauthed requests: `"Page size must be between 1 & 20 for unauthenticated requests."`

File: api/test/unit/controllers/test_search_controller.py (hunk 3 above)
Comment (krysal) [inline api/test/unit/controllers/test_search_controller.py:43]:
```suggestion
```

**Review by zackkrida — APPROVED** [review]
Looks great using the test instructions from the prior PR.

File: api/catalog/api/controllers/search_controller.py (hunk 1 above)
Comment (krysal) [reply api/catalog/api/controllers/search_controller.py:508]: So glad to see this function simplified even more!

**Review by krysal — APPROVED** [review]
Thank you both @sarayourfriend & @dhruvkb for working on this! Looks great 💯

**Pull request merged**